src/Features/CSharp/Portable/ConvertToRecord/ConvertToRecordEngine.cs

```diff
@@ -486,31 +492,52 @@ private static SyntaxList<AttributeListSyntax> GetModifiedAttributeListsForPrope
-                    var newInitializer = objectCreationExpression.Initializer!.RemoveNodes(
-                        expressions
-                            // index can be null if expression is the factory created default or null
-                            .Select(expression => expression.Parent as AssignmentExpressionSyntax)
-                            .WhereNotNull(),
-                        RemovalOptions)!;
-
-                    // if there are no more assignments other than the ones that
-                    // could go in the primary constructor, we can remove the block entirely
-                    if (newInitializer.Expressions.IsEmpty())
-                    {
-                        newInitializer = null;
-                    }
+                    var expressionIndices = expressions.SelectAsArray(
+                        // if initializer was null we wouldn't have found expressions
+                        // any constructed nodes (default/null) should give -1 because parent is null
+                        expression => objectCreationExpression.Initializer!.Expressions.IndexOf(expression.Parent));
 
-                    // replace: new C { Foo = 0; Bar = false; };
-                    // with: new C(0, false);
-                    var replacementNode = SyntaxFactory.ObjectCreationExpression(
-                        objectCreationExpression.NewKeyword,
-                        objectCreationExpression.Type.WithoutTrailingTrivia(),
-                        SyntaxFactory.ArgumentList(SyntaxFactory.SeparatedList(
-                            expressions.Select(expression => SyntaxFactory.Argument(expression.WithoutTrivia())))),
-                        newInitializer);
+                    documentEditor.ReplaceNode(objectCreationExpression, (node, generator) =>
+                    {
+                        var updatedObjectCreation = (ObjectCreationExpressionSyntax)node;
+                        var newInitializer = (InitializerExpressionSyntax)generator
+                            .RemoveNodes(updatedObjectCreation.Initializer!,
+                                expressionIndices
+                                    .Where(i => i != -1)
+                                    .Select(i => updatedObjectCreation.Initializer!.Expressions[i]));
+
+                        // if there are no more assignments other than the ones that
+                        // could go in the primary constructor, we can remove the block entirely
+                        if (newInitializer.Expressions.IsEmpty())
+                        {
+                            newInitializer = null;
+                        }
 
-                    documentEditor.ReplaceNode(objectCreationExpression, replacementNode);
+                        var updatedExpressions = expressions.Select((expression, idx) =>
+                        {
+                            if (expression.Parent == null)
+                            {
```

Comment:
A bit confused as to this condition here?

Reply:
Yes, you're right, it would have had an index of -1. I realize now we can just zip the expressions and indices together to make it a little cleaner.